src/main/java/com/google/api/codegen/ProtocGeneratorMain.java

```diff
@@ -0,0 +1,125 @@
+import java.util.List;
+
+/**
+ * Entrypoint for protoc-plugin invoked generation. Protoc passes input via std.in as a serialized
+ * CodeGeneratorRequest, and expects to read a CodeGeneratorResponse from std.out.
+ */
+public class ProtocGeneratorMain {
+
+  private static final ArtifactType DEFAULT_ARTIFACT_TYPE = GAPIC_CODE;
+
+  public static void main(String[] args) {
+    // TODO(andrealin): Implement. Handle I/O interfacing with protoc.
+  }
+
+  @VisibleForTesting
+  // Parses the InputStream for a CodeGeneratorRequest and returns the generated output in a
+  // CodeGeneratorResponse.
+  public static CodeGeneratorResponse generate(CodeGeneratorRequest request) {
+    try {
+      ToolOptions toolOptions = parseOptions(request);
+
+      GapicGeneratorApp codeGen = new GapicGeneratorApp(toolOptions, DEFAULT_ARTIFACT_TYPE, true);
+
+      codeGen.run();
+      CodeGeneratorResponse response = codeGen.getCodeGeneratorProtoResponse();
+      if (response == null) {
+        throw new RuntimeException(collectDiags(codeGen));
+      }
+      return response;
+    } catch (Exception e) {
+      StringWriter sw = new StringWriter();
+      PrintWriter pw = new PrintWriter(sw);
+      e.printStackTrace(pw);
+      pw.flush();
+      return PluginProtos.CodeGeneratorResponse.newBuilder().setError(sw.toString()).build();
+    }
+  }
+
+  private static ToolOptions parseOptions(CodeGeneratorRequest request) throws Exception {
+    List<FileDescriptorProto> fileDescriptorProtoList = request.getProtoFileList();
+    FileDescriptorSet descriptorSet =
+        FileDescriptorSet.newBuilder().addAllFile(fileDescriptorProtoList).build();
+
+    // Write out DescriptorSet to temp file.
+    File descriptorSetFile;
+
+    descriptorSetFile = File.createTempFile("api", ".desc");
+    FileOutputStream fileoutput = new FileOutputStream(descriptorSetFile);
+    descriptorSet.writeTo(fileoutput);
+    fileoutput.close();
+    descriptorSetFile.deleteOnExit();
+
+    List<String> parsedArgs = new LinkedList<>();
+    parsedArgs.add("--descriptor_set");
+    parsedArgs.add(descriptorSetFile.getAbsolutePath());
+
+    // For now, assume there will only be one proto package to be generated.
+    String firstFiletoGenerate = request.getFileToGenerate(0);
+    String protoPackage =
+        request
+            .getProtoFileList()
+            .stream()
+            .filter(f -> f.getName().equals(firstFiletoGenerate))
+            .findAny()
+            .get()
+            .getPackage();
+    parsedArgs.add("--package");
+    parsedArgs.add(protoPackage);
+
+    // Parse plugin params, ignoring unknown params.
+    String[] requestArgs = request.getParameter().split(",");
+    for (String arg : requestArgs) {
+      if (Strings.isNullOrEmpty(arg)) continue;
+      parsedArgs.add("--" + arg);
+    }
+
+    String[] argsArray = parsedArgs.toArray(new String[] {});
+
+    return GeneratorMain.createCodeGeneratorOptionsFromProtoc(argsArray);
```

Comment:
This parseOptions function seems deeply tied together with the createCodeGeneratorOptionsFromProtoc function. Why does the createCodeGeneratorOptionsFromProtoc function belong in a different class?

Reply:
The `createCodeGeneratorOptionsFromProtoc` looks like all the other command-line-options-processing methods in `GeneratorMain`, plus it uses the `static final Option` fields defined in `GeneratorMain`. I figure it's easier to figure out and compare/contrast what that method is doing when placed next to other very similar methods.

Reply:
OK, sgtm. Perhaps we could consider a dedicated parsing class that both ProtocGeneratorMain and GeneratorMain share? But I haven't thought that through enough, and doesn't need to be in this PR.